### PR TITLE
Set Service Principal in azurejson secret when AzureCluster has IdentityRef but not using Managed Identity

### LINKED ISSUE #1517: azurejson controller doesn't have access to ClientID/Secret/TenantID when using AzureClusterIdentity
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

**What did you expect to happen:** The azure.json controller doesn’t have access to the Client secret/id/tenant ID when using AzureClusterIdentity and the manager creds aren’t set. This causes the azurejson secrets to be generated without aadClientID and aadClientSecret set, when not using managed identity, which causes cloud provider to fail (kube-controller-manager goes into crashloop).

Workaround is to keep setting manager credentials (base 64 encoded env vars), or to use managed identity.

Potential fixes to consider (just spitballing):

- Find a way to give access to the creds in the controller through aad-pod-identity
- create an identity for the machine (?)
- make managed system identity default and require env vars for SP cloud provider identity *but this requires the AzureClusterIdentity SP to have Owner access*
- ???

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**

**What this PR does / why we need it**: The azure.json controller didn't have access to the Client secret/id/tenant ID when using AzureClusterIdentity and the manager creds were not set. This caused the azurejson secrets to be generated without aadClientID and aadClientSecret set, when not using managed identity, which causes cloud provider to fail (kube-controller-manager goes into crashloop).

Also adds a warning when not using Managed Identity since it is less secure and requires writing creds to disk.

To ensure the integrity of these changes, the tilt, e2e and conformance scripts have been modified to not set base64 encoded variables so that the manager credentials are no longer set (this is what prevented us from catching this bug in the first place). Intentionally left ci-entrypoint untouched, will follow up on that one later.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1517

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Get Cloud Provider config Service Principal credentials from AzureClusterIdentity when not using Managed Identity 
```

### CODE REVIEW

**Comment by nader-ziada** [comment]
lint errors
```
azure/scope/identity.go:22: File is not `goimports`-ed (goimports)
	"k8s.io/apimachinery/pkg/types"
azure/scope/identity.go:250:1: comment on exported method `AzureCredentialsProvider.GetTenantID` should be of the form `GetTenantID ...` (golint)
// GetClientID returns the Tenant ID associated with the AzureCredentialsProvider's Identity.
^
```

**Comment by CecileRobertMichon** [comment]
I expect the private cluster and self-hosted specs to fail because of known Calico issue identified in #1360 but running all the tests to make sure nothing else fails before I re-add the env vars for those two

**Comment by CecileRobertMichon** [comment]
/assign @devigned @nader-ziada @shysank 

ready for review

**Comment by nader-ziada** [comment]
great PR, will just wait and see what happens with the tests

File: controllers/azurejson_machinetemplate_controller.go [hunk 1]
```diff
@@ -145,6 +146,14 @@ func (r *AzureJSONTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		userAssignedIdentityIfExists = azureMachineTemplate.Spec.Template.Spec.UserAssignedIdentities[0].ProviderID
 	}
 
+	if azureMachineTemplate.Spec.Template.Spec.Identity == infrav1.VMIdentityNone {
+		warningMessage := "You are using Service Principal authentication for Cloud Provider Azure which is less secure than Managed Identity."
```
Comment (devigned) [inline controllers/azurejson_machinetemplate_controller.go:150]: Would you consider moving this into a const?

**Review by devigned — CHANGES_REQUESTED** [review]
Just a consideration request for moving the warning message into a const. Outside of that, the pr lgtm.

**Comment by nader-ziada** [comment]
The two failures here: `pull-cluster-api-provider-azure-e2e-full` don't seem to be related to this PR, the `GPU` and `AKS identity` have been know to have flakes

File: azure/scope/identity.go [hunk 2]
```diff
@@ -219,6 +226,33 @@ func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, resourceMa
...
 }
 
+// GetClientID returns the Client ID associated with the AzureCredentialsProvider's Identity.
+func (p *AzureCredentialsProvider) GetClientID() string {
+	return p.Identity.Spec.ClientID
+}
+
+// GetClientSecret returns the Client Secret associated with the AzureCredentialsProvider's Identity.
+// NOTE: this only works if the Identity references a Service Principal Client Secret.
+// If using another type of credentials, such a Certificate, we return an empty string.
```
Comment (shysank) [inline azure/scope/identity.go:236]: Clarification: How does cloud provider get the credentials when using other type of credentials (eg. certificate)? Is it expected to use a managed identity for such cases?

**Review by shysank — COMMENTED** [review]
lgtm with one clarification

File: azure/scope/identity.go (hunk 2 above)
Comment (nader-ziada) [reply azure/scope/identity.go:236]: The answer to your question is coming here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1521

**Review by devigned — APPROVED** [review]
/lgtm

**Pull request merged**